hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ListIterator.java

```diff
@@ -0,0 +1,352 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.om;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.hdds.utils.IOUtils;
+import org.apache.hadoop.hdds.utils.db.CopyObject;
+import org.apache.hadoop.hdds.utils.db.Table;
+import org.apache.hadoop.hdds.utils.db.TableIterator;
+import org.apache.hadoop.hdds.utils.db.cache.CacheKey;
+import org.apache.hadoop.hdds.utils.db.cache.CacheValue;
+import org.apache.hadoop.ozone.om.helpers.BucketLayout;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.PriorityQueue;
+import java.util.TreeMap;
+
+import static org.apache.hadoop.ozone.om.lock.OzoneManagerLock.Resource.BUCKET_LOCK;
+
+/**
+ * Common class to do listing of resources after merging
+ * rocksDB table cache & actual table.
+ */
+public class ListIterator {
+
+  /**
+   * Interface for iteration of Heap Entries.
+   */
+  public interface ClosableIterator extends Iterator<HeapEntry>, Closeable {
+
+  }
+
+  /**
+   * Entry to be added to the heap.
+   */
+  public static class HeapEntry implements Comparable<HeapEntry> {
+    private final int entryIteratorId;
+    private final String tableName;
+    private final String key;
+    private final Object value;
+
+    HeapEntry(int entryIteratorId, String tableName, String key,
+              Object value) {
+      this.entryIteratorId = entryIteratorId;
+      this.tableName = tableName;
+      this.key = key;
+      this.value = value;
+    }
+
+    public String getKey() {
+      return this.key;
+    }
+
+    private int getEntryIteratorId() {
+      return this.entryIteratorId;
+    }
+
+    public String getTableName() {
+      return tableName;
+    }
+
+    public Object getValue() {
+      return value;
+    }
+
+    public int compareTo(HeapEntry other) {
+      return Comparator.comparing(HeapEntry::getKey)
+          .thenComparing(HeapEntry::getEntryIteratorId).compare(this, other);
```

Comment:
Why do we also need to compare `entryIteratorId`? I'm not getting a grasp of the ordering here.

Reply:
Getting the entry from cache directly